File: wasm/src/voice-clone.ts
/**
 * Captures the short reference clip that zero-shot voice cloning needs.
 *
 * ```ts
 * const clone = tts.startCloning();
 * clone.onReady(() => status.textContent = 'Got it — you can stop talking.');
 * await clone.fromMicrophone();
 * await tts.cloneFrom(clone);
 * ```
 *
 * Finding a usable clip means locating a window of the recording that is mostly
 * speech rather than silence or breathing. That search runs in the core (see
 * `moonshine_extract_speech_clip`), so the browser, iOS and Android bindings
 * all agree on what a good clip looks like. Extract stays VAD-only; ZipVoice
 * clone ASR refine + transcript happen once inside {@link TextToSpeech.cloneFrom}.
 */

import { resampleTo16k } from './mic-transcriber.js';
import { loadMoonshineModule, type MoonshineModule } from './module.js';

const TARGET_SAMPLE_RATE = 16000;
/** How much new audio to accumulate between speech searches. */
const SEARCH_INTERVAL_SECONDS = 0.25;
/** Give up looking for a good window after this much recording. */
const DEFAULT_MAX_RECORD_SECONDS = 20;

export interface VoiceCloneOptions {
  /** Length of the reference clip in seconds. Defaults to 4. */
  clipSeconds?: number;
  /** How much of the clip has to be speech. Defaults to 2. */
  minimumSpeechSeconds?: number;
}

export class VoiceClone {
  private readonly module: MoonshineModule;
  private readonly ttsHandle: number;
  private readonly clipSeconds: number;
  private readonly minimumSpeechSeconds: number;

  private chunks: Float32Array[] = [];
  private sampleCount = 0;
  private samplesSinceSearch = 0;
  private clip?: Float32Array;
  private clipTranscript?: string;
  private speech = 0;
  private readyCallbacks: Array<() => void> = [];
  private progressCallbacks: Array<
    (recordedSeconds: number, speechSeconds: number) => void
  > = [];
  private stopCapture?: () => Promise<void>;

  constructor(
    module: MoonshineModule,
    ttsHandle: number,
    options: VoiceCloneOptions = {},
  ) {
    this.module = module;
    this.ttsHandle = ttsHandle;
    this.clipSeconds = options.clipSeconds ?? 4;
    this.minimumSpeechSeconds = options.minimumSpeechSeconds ?? 2;
    if (!module.extractSpeechClip) {
      throw new Error(
        'This Moonshine WASM build does not support speech clip extraction.',
      );
    }
  }

  /** Fires once, as soon as enough speech has been captured. */
  onReady(callback: () => void): this {
    if (this.clip) {
      callback();
    } else {
      this.readyCallbacks.push(callback);
    }
    return this;
  }

  /** Reports how long the caller has been recording and how much was speech. */
  onProgress(
    callback: (recordedSeconds: number, speechSeconds: number) => void,
  ): this {
    this.progressCallbacks.push(callback);
    return this;
  }

  /** True once {@link audio} holds a usable reference clip. */
  get isReady(): boolean {
    return this.clip !== undefined;
  }

  /** Speech found in the best window so far, in seconds. */
  get speechSeconds(): number {
    return this.speech;
  }

  /** Transcript is unused for VAD capture; cloneFrom fills it via create-time ASR. */
  get transcript(): string | undefined {
    return this.clipTranscript;
  }

  get recordedSeconds(): number {
    return this.sampleCount / TARGET_SAMPLE_RATE;
  }

  /** The captured clip (16 kHz mono), or `undefined` until {@link isReady}. */
  get audio(): Float32Array | undefined {
    return this.clip;
  }

  get sampleRate(): number {
    return TARGET_SAMPLE_RATE;
  }

  /**
   * Feeds captured audio in. Call this from your own audio pipeline; the search
   * for a usable window runs a few times a second rather than on every chunk.
   */
  addAudio(pcm: Float32Array, sampleRate: number): void {
    if (this.clip || pcm.length === 0) return;
    const resampled = resampleTo16k(pcm, sampleRate);
    this.chunks.push(resampled);
    this.sampleCount += resampled.length;
    this.samplesSinceSearch += resampled.length;
    if (
      this.samplesSinceSearch <
      SEARCH_INTERVAL_SECONDS * TARGET_SAMPLE_RATE
    ) {
      return;
    }
    this.samplesSinceSearch = 0;
    this.search();
  }

  /**
   * Opens the microphone and records until there is enough speech, or until
   * `maxSeconds` have passed. Resolves with the clip, which is also available
   * from {@link audio}.
   */
  async fromMicrophone(
    options: {
      maxSeconds?: number;
      audioConstraints?: MediaTrackConstraints | boolean;
    } = {},
  ): Promise<Float32Array> {
    const maxSeconds = options.maxSeconds ?? DEFAULT_MAX_RECORD_SECONDS;
    const mediaStream = await navigator.mediaDevices.getUserMedia({
      audio: options.audioConstraints ?? true,
    });
    const context = new AudioContext();
    // A context created without user activation starts suspended, and never
    // pulls audio: recording would appear to run and capture nothing but
    // silence until maxSeconds expired.
    if (context.state === 'suspended') await context.resume();
    const inputRate = context.sampleRate;
    const source = context.createMediaStreamSource(mediaStream);
    const node = context.createScriptProcessor(4096, 1, 1);

    let finish: (clip: Float32Array) => void;
    let fail: (err: Error) => void;
    const done = new Promise<Float32Array>((resolve, reject) => {
      finish = resolve;
      fail = reject;
    });

    this.stopCapture = async () => {
      node.onaudioprocess = null;
      node.disconnect();
      source.disconnect();
      mediaStream.getTracks().forEach((t) => t.stop());
      await context.close();
      this.stopCapture = undefined;
    };

    node.onaudioprocess = (event) => {
      this.addAudio(
        new Float32Array(event.inputBuffer.getChannelData(0)),
        inputRate,
      );
      if (this.clip) {
        finish(this.clip);
      } else if (this.recordedSeconds >= maxSeconds) {
        // Out of patience: take the best window we have, even a quiet one.
        this.search({ acceptAnything: true });
        if (this.clip) {
          finish(this.clip);
        } else {
          fail(
            new Error(
              `No speech detected in ${maxSeconds}s of recording. Try again somewhere quieter.`,
            ),
          );
        }
      }
    };

    source.connect(node);
    node.connect(context.destination);

    try {
      return await done;
    } finally {
      await this.stopCapture?.();
    }
  }

  /** Stops an in-flight {@link fromMicrophone} capture. */
  async cancel(): Promise<void> {
    await this.stopCapture?.();
  }

  /** Throws away everything captured so far. */
  reset(): void {
    this.chunks = [];
    this.sampleCount = 0;
    this.samplesSinceSearch = 0;
    this.clip = undefined;
    this.clipTranscript = undefined;
    this.speech = 0;
  }

  private search(options: { acceptAnything?: boolean } = {}): void {
    const recording = concat(this.chunks, this.sampleCount);
    const result = this.module.extractSpeechClip!(
      recording,
      TARGET_SAMPLE_RATE,
      this.ttsHandle,
      this.clipSeconds,
      options.acceptAnything ? 0 : this.minimumSpeechSeconds,
    );
    this.speech = result.speechDuration;
    for (const cb of this.progressCallbacks) {
      cb(this.recordedSeconds, this.speech);
    }
    if (result.isComplete && result.audio && result.audio.length > 0) {
      this.clip = result.audio;
      this.clipTranscript = result.transcript || undefined;
      const callbacks = this.readyCallbacks;
      this.readyCallbacks = [];
      for (const cb of callbacks) cb();
    }
  }
}

/**
 * Pulls a reference clip out of already-recorded audio, without any capture
 * loop. Used by `TextToSpeech.cloneFrom` when given a file or buffer.
 */
export async function extractSpeechClip(
  audio: Float32Array,
  sampleRate: number,
  options: VoiceCloneOptions & {
    module?: MoonshineModule;
    ttsHandle: number;
  },
): Promise<Float32Array> {
  const module = options.module ?? (await loadMoonshineModule());
  if (!module.extractSpeechClip) {
    throw new Error(
      'This Moonshine WASM build does not support speech clip extraction.',
    );
  }
  const clipSeconds = options.clipSeconds ?? 4;
  const result = module.extractSpeechClip(
    audio,
    sampleRate,
    options.ttsHandle,
    clipSeconds,
    options.minimumSpeechSeconds ?? 2,
  );
  if (result.isComplete && result.audio && result.audio.length > 0) {
    return result.audio;
  }
  // Nothing clearly speech-like. Rather than fail outright, fall back to the
  // best window the detector found, or the start of the recording if it found
  // nothing at all — a poor clone beats no clone for a caller who explicitly
  // handed us this audio.
  const resampled = resampleTo16k(audio, sampleRate);
  const wanted = Math.round(clipSeconds * TARGET_SAMPLE_RATE);
  if (resampled.length <= wanted) return resampled;
  const from = Math.min(
    Math.max(0, Math.round(result.startTime * TARGET_SAMPLE_RATE)),
    resampled.length - wanted,
  );
  return resampled.slice(from, from + wanted);
}

function concat(chunks: Float32Array[], total: number): Float32Array {
  const out = new Float32Array(total);
  let offset = 0;
  for (const chunk of chunks) {
    out.set(chunk, offset);
    offset += chunk.length;
  }
  return out;
}

File: language-bindings/android/java/main/java/ai/moonshine/voice/VoiceClone.java
package ai.moonshine.voice;

import android.content.Context;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Captures the short reference clip that zero-shot voice cloning needs.
 *
 * <pre>{@code
 * VoiceClone clone = tts.startCloning();
 * clone.onReady(() -> status.setText("Got it - you can stop talking."));
 * clone.fromMicrophone();
 * tts.cloneFrom(clone);
 * }</pre>
 *
 * <p>Finding a usable clip means locating a window of the recording that is
 * mostly speech rather than silence or breathing. That search runs in the core,
 * so the browser, iOS and Android bindings all agree on what a good clip looks
 * like. No model download is involved: the voice-activity detector is compiled
 * into the library.
 */
public final class VoiceClone {

    /** Sample rate of the clip handed back by {@link #getAudio()}. */
    public static final int CLIP_SAMPLE_RATE = SpeechClip.SAMPLE_RATE;
    /** How long {@link #fromMicrophone()} keeps listening before giving up. */
    public static final float DEFAULT_MAX_RECORD_SECONDS = 20f;
    /** How much new audio to accumulate between speech searches. */
    private static final float SEARCH_INTERVAL_SECONDS = 0.25f;

    /** Fires once, as soon as enough speech has been captured. */
    public interface ReadyCallback {
        void onReady();
    }

    /** Reports seconds recorded so far and seconds of speech in the best window. */
    public interface CloneProgressCallback {
        void onProgress(float recordedSeconds, float speechSeconds);
    }

    private final Context appContext;
    private final int ttsHandle;
    private final float clipDurationSeconds;
    private final float minimumSpeechSeconds;

    private final Object lock = new Object();
    private final List<float[]> recording = new ArrayList<>();
    private int recordedSamples = 0;
    private int recordingSampleRate = CLIP_SAMPLE_RATE;
    private int samplesSinceSearch = 0;
    @Nullable private float[] clip;
    @Nullable private String transcript;
    private float speechSeconds = 0;

    private final CopyOnWriteArrayList<ReadyCallback> readyCallbacks = new CopyOnWriteArrayList<>();
    private final CopyOnWriteArrayList<CloneProgressCallback> progressCallbacks =
            new CopyOnWriteArrayList<>();

    private volatile boolean capturing = false;

    public VoiceClone(Context context, int ttsHandle) {
        this(context, ttsHandle, 4f, 2f);
    }

    public VoiceClone(Context context, int ttsHandle, float clipDurationSeconds,
            float minimumSpeechSeconds) {
        if (context == null) {
            throw new IllegalArgumentException("context is required");
        }
        JNI.ensureLibraryLoaded();
        this.appContext = context.getApplicationContext();
        this.ttsHandle = ttsHandle;
        this.clipDurationSeconds = clipDurationSeconds;
        this.minimumSpeechSeconds = minimumSpeechSeconds;
    }

    /** Fires once enough speech has been captured, or immediately if it already has. */
    public VoiceClone onReady(ReadyCallback callback) {
        boolean alreadyReady;
        synchronized (lock) {
            alreadyReady = clip != null;
            if (!alreadyReady) {
                readyCallbacks.add(callback);
            }
        }
        if (alreadyReady) {
            callback.onReady();
        }
        return this;
    }

    /** Reports recording and speech durations so the app can show progress. */
    public VoiceClone onProgress(CloneProgressCallback callback) {
        progressCallbacks.add(callback);
        return this;
    }

    /** True once {@link #getAudio()} holds a usable reference clip. */
    public boolean isReady() {
        synchronized (lock) {
            return clip != null;
        }
    }

    /** The captured clip (16 kHz mono), or null until {@link #isReady()}. */
    @Nullable
    public float[] getAudio() {
        synchronized (lock) {
            return clip;
        }
    }

    public int getSampleRate() {
        return CLIP_SAMPLE_RATE;
    }

    /** Speech found in the best window so far, in seconds. */
    public float getSpeechSeconds() {
        synchronized (lock) {
            return speechSeconds;
        }
    }

    /** Unused for VAD capture; cloneFrom fills the transcript via create-time ASR. */
    @Nullable
    public String getTranscript() {
        synchronized (lock) {
            return transcript;
        }
    }

    public float getRecordedSeconds() {
        synchronized (lock) {
            return recordingSampleRate > 0 ? (float) recordedSamples / recordingSampleRate : 0f;
        }
    }

    /**
     * Feeds captured audio in. Call this from your own audio pipeline; the
     * search for a usable window runs a few times a second rather than on every
     * chunk.
     */
    public void addAudio(float[] pcm, int sampleRate) {
        boolean due;
        synchronized (lock) {
            if (clip != null || pcm == null || pcm.length == 0 || sampleRate <= 0) {
                return;
            }
            if (sampleRate != recordingSampleRate) {
                // Mixed rates in one buffer would make the clip come out at the
                // wrong speed, so a change starts the recording over.
                recording.clear();
                recordedSamples = 0;
                recordingSampleRate = sampleRate;
                samplesSinceSearch = 0;
            }
            recording.add(pcm);
            recordedSamples += pcm.length;
            samplesSinceSearch += pcm.length;
            due = samplesSinceSearch >= SEARCH_INTERVAL_SECONDS * sampleRate;
            if (due) {
                samplesSinceSearch = 0;
            }
        }
        if (due) {
            search(false);
        }
    }

    /** Records until there is enough speech, or {@link #DEFAULT_MAX_RECORD_SECONDS} elapse. */
    public float[] fromMicrophone() {
        return fromMicrophone(DEFAULT_MAX_RECORD_SECONDS);
    }

    /**
     * Opens the microphone and records until there is enough speech, or until
     * {@code maxSeconds} have passed. Blocks; call from a background thread.
     *
     * @return the clip, which is also available from {@link #getAudio()}.
     */
    public float[] fromMicrophone(float maxSeconds) {
        MicrophonePermission.ensureGranted(appContext);
        MicCaptureProcessor capture = new MicCaptureProcessor();
        Thread thread = new Thread(capture, "moonshine-clone-capture");
        thread.setDaemon(true);
        capturing = true;
        thread.start();
        try {
            long deadline = System.nanoTime() + (long) (maxSeconds * 1_000_000_000L);
            while (capturing && !isReady()) {
                if (System.nanoTime() >= deadline) {
                    // Out of patience: take the best window we have, even a quiet one.
                    search(true);
                    break;
                }
                addAudio(capture.consumeAudio(), CLIP_SAMPLE_RATE);
                try {
                    Thread.sleep(20);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        } finally {
            capturing = false;
            thread.interrupt();
        }
        float[] audio = getAudio();
        if (audio == null) {
            throw new IllegalStateException("No speech detected in " + (int) maxSeconds
                    + "s of recording. Try again somewhere quieter.");
        }
        return audio;
    }

    /** Stops an in-flight {@link #fromMicrophone(float)} capture. */
    public void cancel() {
        capturing = false;
    }

    /** Throws away everything captured so far. */
    public void reset() {
        synchronized (lock) {
            recording.clear();
            recordedSamples = 0;
            samplesSinceSearch = 0;
            clip = null;
            transcript = null;
            speechSeconds = 0;
        }
    }

    // -- Internals -----------------------------------------------------------

    private void search(boolean acceptAnything) {
        float[] samples;
        int rate;
        synchronized (lock) {
            if (clip != null) {
                return;
            }
            samples = flattenLocked();
            rate = recordingSampleRate;
        }
        if (samples.length == 0) {
            return;
        }
        SpeechClip found = JNI.moonshineExtractSpeechClip(samples, rate, ttsHandle,
                clipDurationSeconds, acceptAnything ? 0f : minimumSpeechSeconds);
        if (found == null) {
            return;
        }

        float recorded = (float) samples.length / rate;
        List<ReadyCallback> ready = null;
        synchronized (lock) {
            speechSeconds = found.speechDuration;
            if (found.audio != null && found.audio.length > 0) {
                clip = found.audio;
                transcript = found.transcript;
                ready = new ArrayList<>(readyCallbacks);
                readyCallbacks.clear();
            }
        }
        for (CloneProgressCallback callback : progressCallbacks) {
            callback.onProgress(recorded, found.speechDuration);
        }
        if (ready != null) {
            for (ReadyCallback callback : ready) {
                callback.onReady();
            }
        }
    }

    private float[] flattenLocked() {
        float[] all = new float[recordedSamples];
        int offset = 0;
        for (float[] chunk : recording) {
            System.arraycopy(chunk, 0, all, offset, chunk.length);
            offset += chunk.length;
        }
        // Coalesce so the next search does not walk a long list of chunks.
        recording.clear();
        if (all.length > 0) {
            recording.add(all);
        }
        return all;
    }
}
